keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,516 @@
+  // A null or empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  MaxRestarts int32 `json:"maxRestarts,omitempty"`
+  // List of failure policy rules for this JobSet.
+  // For a given Job failure, the rules will be evaluated in order,
+  // and only the first matching rule will be executed.
+  Rules []FailurePolicyRule `json:"rules,omitempty"`
+}
+
+```
+
+### Constraints
+
+- For a given Job failure, the rules will be evaluated in order, and only the first matching rule will be executed. This
+follows an established pattern used by PodFailurePolicy in the Job controller.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
+  - `FailedIndexes`
+
+### Implmentation
+
+The core part of the implementation will be defining what specific mechanisms the JobSet controller uses to implement
+the behavior defined for each FailurePolicyAction type:
+
+1) `FailJobSet`: To fail the JobSet immediately without restarting, the controller updates the JobSet status to failed.
+
+2) `Ignore`: To "ignore" the failure (i.e., restart the JobSet without counting it against `MaxRestarts`), the controller
+will delete all child jobs and allow the normal reconciliation process to recreate them. This is in contrast to how
+we restart the JobSet **without** ignoring the failure, which is done by incrementing the counter on the JobSet annotation
+`jobset.sigs.k8s.io/restart-attempt`, which on subsequent reconiliations will trigger job deletion and recreation for any
+child job with a `jobset.sigs.k8s.io/restart-attempt` counter value less than that of the JobSet.
+
+3) `RestartReplicatedJob`: To restart the child jobs of a specific replicated job, the controller will delete the child
+jobs of the target replicated job, **without incrementing the restart attempt annotation**. The jobs will then be 
+recreated via the normal reconciliation process.
+
+4) `RestartJob`: To restart a single child job without restarting the entire JobSet, the controller will delete that 
+particular child job, **without incrementing the restart attempt annotation**, and allow the normal reconciliation
+process to recreate it.
+
+5) `FailJob`: To leave a particular child job in a failed state without restarting it or restarting the JobSet, the
+controller will simply do nothing, taking no action on this job.
+
```

Comment:
I would hold off implementing those three actions, we can add them later. For the first two, we need to think of how to define limits on the number of restarts, for the last one I think we need to mark the job as failed forever just like FailIndex in the Job API, and to do that, we need to use finalizers, which will complicate the implementation.

Perhaps move them to a future work section and leave the above points as open questions for discussion.

Reply:
That makes sense, especially since 2 of the 3 rely on an upstream dependency which in the best case scenario, will only be in alpha in 1.30. 

Moved these to a "Future Work" section at the bottom, explaining the unresolved ambiguity and the upstream dependencies under review.